File: src/backend/DealMatcher.Backend/src/DealMatcher.Backend.Core/Aggregates/Conversation/Conversation.cs
namespace DealMatcher.Backend.Core.Aggregates.Conversation;

public class Conversation : DealMatcherEntityBase,
    IAggregateRoot
{
    public int OfferId { get; private set; }
    public int BuyerId { get; }
    public int SellerId { get; }
    public Offer.Offer Offer { get; private set; } = null!;
    public User.User Buyer { get; private set; } = null!;
    public User.User Seller { get; private set; } = null!;

    public string LastMessage { get; private set; }
    public DateTime LastMessageAt { get; private set; }
    public int UnreadCount { get; private set; }
    public ConversationStatus Status { get; private set; }
    public List<Message> Messages { get; } = [];

    public Conversation(
        int offerId,
        int buyerId,
        int sellerId,
        string initialMessage
    )
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(offerId);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(buyerId);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(sellerId);

        OfferId = offerId;
        BuyerId = buyerId;
        SellerId = sellerId;

        LastMessage = initialMessage;
        LastMessageAt = DateTime.UtcNow;

        Status = ConversationStatus.Active;

        Messages = [];
        AddMessage(buyerId, initialMessage);
    }

#pragma warning disable CS8618
    private Conversation()
    {
        /* EF */
    }
#pragma warning restore CS8618

    public Message AddMessage(int senderId, string content)
    {
        if (Status == ConversationStatus.Closed)
        {
            throw new InvalidOperationException("Cannot add new message to closed conversation");
        }

        if (senderId != BuyerId && senderId != SellerId)
            throw new ArgumentException(
                "Sender must belong to conversation.");

        ArgumentException.ThrowIfNullOrWhiteSpace(content);

        var message = new Message(Id, senderId, content.Trim());

        Messages.Add(message);

        LastMessage = message.Content;
        LastMessageAt = message.CreatedAt;

        UnreadCount++;

        return message;
    }

    public void MarkAsRead()
    {
        UnreadCount = 0;
    }

    public void CloseConversation()
    {
        Status = ConversationStatus.Closed;
    }
}

File: src/frontend/DealMatcher.Frontend/lib/Models/conversation.dart
import 'package:frontend/Models/offer.dart';

class Conversation {
  final int _id;
  final int _offerId;
  final Offer _offer;
  final int _buyerId;
  final int _sellerId;
  final String _buyerName;
  final String _sellerName;
  final String _lastMessage;
  final DateTime _lastMessageAt;
  final int _unreadCount;
  final String _status;
  final DateTime _createdAt;

  int get id => _id;
  int get offerId => _offerId;
  Offer get offer => _offer;
  int get buyerId => _buyerId;
  int get sellerId => _sellerId;
  String get buyerName => _buyerName;
  String get sellerName => _sellerName;
  String get lastMessage => _lastMessage;
  DateTime get lastMessageAt => _lastMessageAt;
  int get unreadCount => _unreadCount;
  String get status => _status;
  DateTime get createdAt => _createdAt;

  Conversation.fromJson(Map<String, dynamic> json)
    : _id = json['id'] ?? 0,
      _offer = Offer.fromJson(json['offer'] as Map<String, dynamic>),
      _offerId = json['offer']?['id'] ?? json['offerId'] ?? 0,
      _buyerId = json['buyer']?['id'] ?? json['buyerId'] ?? 0,
      _sellerId = json['seller']?['id'] ?? json['sellerId'] ?? 0,
      _buyerName = json['buyer']?['name'] ?? '',
      _sellerName = json['seller']?['name'] ?? '',
      _lastMessage = json['lastMessage'] ?? '',
      _lastMessageAt =
          DateTime.tryParse(json['lastMessageAt'] ?? '') ?? DateTime.now(),
      _unreadCount = json['unreadCount'] ?? 0,
      _status = json['status'] ?? 'CLOSED',
      _createdAt = DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime.now();
}
